File: Day #1175/solution.go
// Day 1175: Maximum subarray sum in a circular array (empty allowed -> 0).
// Answer = max(0, kadaneMax, total - kadaneMin); total-min covers the wrap case.
// Time O(N), Space O(1).
package main

import "fmt"

func maxCircularSubarray(a []int) int {
	total, curMax, bestMax, curMin, bestMin := 0, 0, 0, 0, 0
	max := func(x, y int) int { if x > y { return x }; return y }
	min := func(x, y int) int { if x < y { return x }; return y }
	for _, x := range a {
		total += x
		curMax = max(x, curMax+x)
		bestMax = max(bestMax, curMax)
		curMin = min(x, curMin+x)
		bestMin = min(bestMin, curMin)
	}
	return max(0, max(bestMax, total-bestMin))
}

func main() {
	fmt.Println(maxCircularSubarray([]int{8, -1, 3, 4})) // 15
	fmt.Println(maxCircularSubarray([]int{-4, 5, 1, 0})) // 6
}

File: Day #1175/solution.py
# Day 1175: Maximum subarray sum in a circular array (empty allowed -> 0).
# Answer = max(0, kadaneMax, total - kadaneMin); total-min covers the wrap case.
# Time O(N), Space O(1).


def max_circular_subarray(a):
    total = cur_max = best_max = cur_min = best_min = 0
    for x in a:
        total += x
        cur_max = max(x, cur_max + x)
        best_max = max(best_max, cur_max)
        cur_min = min(x, cur_min + x)
        best_min = min(best_min, cur_min)
    return max(0, best_max, total - best_min)


if __name__ == "__main__":
    print(max_circular_subarray([8, -1, 3, 4]))  # 15
    print(max_circular_subarray([-4, 5, 1, 0]))  # 6
